predicators/datasets/vlm_input_data.py

```diff
@@ -0,0 +1,243 @@
+        assert len(dummy_goal_feats) == 1
+        assert dummy_goal_feats[0] == "goal_true"
+        curr_task_goal_atom = GroundAtom(
+            goal_predicate, [curr_task_obj_name_to_obj["dummy_goal_obj"]])
+        assert not curr_task_goal_atom.holds(train_task.init)
+        curr_goal_state = train_task.init.copy()
+        curr_goal_state.set(curr_task_obj_name_to_obj["dummy_goal_obj"],
+                            "goal_true", 1.0)
+        assert curr_task_goal_atom.holds(curr_goal_state)
+        goal_states.append(curr_goal_state)
+    return goal_states
+
+
+def _convert_ground_option_trajs_into_lowleveltrajs(
+        option_trajs: List[List[_Option]], dummy_goal_states: List[State],
+        train_tasks: List[Task]) -> List[LowLevelTrajectory]:
+    """Convert option trajectories into LowLevelTrajectories to be used in
+    constructing a Dataset."""
+    assert len(option_trajs) == len(dummy_goal_states) == len(train_tasks)
+    # NOTE: In this LowLevelTrajectory, we assume the low level states
+    # are the same as the init state until the final state.
+    trajs = []
+    for traj_num in range(len(option_trajs)):
+        traj_init_state = train_tasks[traj_num].init
+        curr_traj_states = []
+        curr_trajactions = []
+        for idx_within_traj in range(len(option_trajs[traj_num])):
+            curr_traj_states.append(traj_init_state)
+            curr_trajactions.append(
+                Action(np.zeros(0, dtype=float),
+                       option_trajs[traj_num][idx_within_traj]))
+        # Now, we need to append the final state because there are 1 more
+        # states than actions.
+        curr_traj_states.append(dummy_goal_states[traj_num])
+        curr_traj = LowLevelTrajectory(curr_traj_states, curr_trajactions,
+                                       True, traj_num)
+        trajs.append(curr_traj)
+    return trajs
+
+
+def _pretty_print_atoms_trajs(
+        ground_atoms_trajs: List[List[Set[GroundAtom]]]) -> None:
+    """Debug log the changes in atoms trajectories for easy human-checking."""
+    # Log trajectory information in a very easy to parse format for
+    # debugging.
+    for traj in ground_atoms_trajs:
+        logging.debug(f"Step 0 atoms: {sorted(traj[0])}")
+        for i in range(1, len(traj)):
+            logging.debug(f"Step {i} add effs: {sorted(traj[i] - traj[i-1])}")
+            logging.debug(f"Step {i} del effs: {sorted(traj[i-1] - traj[i])}")
+        logging.debug("\n")
+
+
+def create_ground_atom_data_from_labeled_txt(
+        env: BaseEnv, train_tasks: List[Task],
+        known_options: Set[ParameterizedOption]) -> Dataset:
+    """Given a txt file containing trajectories labelled with VLM predicate
+    values, construct a dataset that can be passed to the rest of our learning
+    pipeline."""
+    dataset_fpath = os.path.join(CFG.data_dir, CFG.handmade_demo_filename)
+    # First, parse this dataset into a structured form.
```

Comment:
after reading the parsing, I'm undecided, will leave it up to you

Reply:
So the real reason for this whole structured thing is that it allows good code sharing between all the VLM parsing code that's about to be in the next PR, and the code that parses these txt files. Basically, the flow looks like this:
**txt files**
input txt -> structured dicts -> regular data structures
**vlm**
input image trajs -> structured dicts -> regular data structures
So we get to share all the code that does `structured dicts -> regular data structures`. What's more, the mapping between input txt and structured dicts is reversible (i.e., we can easily recover the input txt from the structured txt). This will let us dump out txt's from the VLM part of the pipeline, which is super helpful!